### PR TITLE
add boot mode management

### PR DESCRIPTION
**Pull request by dhellmann**
This PR replaces #469 and implements
https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/implicit-boot-mode.md
and https://github.com/metal3-io/metal3-docs/pull/121

### CODE REVIEW

**Comment by dhellmann** [comment]
This is ready for review.

/cc @hardys

File: pkg/provisioner/ironic/ironic.go [hunk 1]
```diff
@@ -203,6 +209,14 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
 	}
 }
 
+// BootMode returns the boot method to be used for the host, using the
+// explicit value if given in the host and falling back to the default
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go]: Fixed

File: pkg/bmc/access.go [hunk 2]
```diff
@@ -49,6 +51,10 @@ type AccessDetails interface {
 	// (such as the kernel and ramdisk locations).
 	DriverInfo(bmcCreds Credentials) map[string]interface{}
 
+	// BootMode returns the default boot mode that should be used for
+	// the host associated with the management controller.
+	BootMode() metal3v1alpha1.BootMode
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/bmc/access.go]: Yes, changed

File: pkg/provisioner/demo/demo.go [hunk 3]
```diff
@@ -303,3 +303,10 @@ func (p *demoProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	// return result, nil
 }
+
+// BootMode returns the boot method to be used for the host, using the
+// explicit value if given in the host and falling back to the default
+// from the BMC driver otherwise.
```
Comment (zaneb) [inline pkg/provisioner/demo/demo.go:309]: This comment could be cleaned up as well.

File: pkg/provisioner/fixture/fixture.go [hunk 4]
```diff
@@ -236,3 +236,10 @@ func (p *fixtureProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// BootMode returns the boot method to be used for the host, using the
+// explicit value if given in the host and falling back to the default
+// from the BMC driver otherwise.
```
Comment (zaneb) [inline pkg/provisioner/fixture/fixture.go:242]: ditto

File: pkg/provisioner/ironic/ironic.go [hunk 5]
```diff
@@ -785,12 +797,33 @@ func (p *ironicProvisioner) getUpdateOptsForNode(ironicNode *nodes.Node) (update
 		},
 	)
 
+	// boot_mode
+	_, ok := ironicNode.InstanceInfo["deploy_boot_mode"]
+	if !ok {
+		op = nodes.AddOp
+	} else {
+		op = nodes.ReplaceOp
+	}
```
Comment (zaneb) [inline pkg/provisioner/ironic/ironic.go:806]: What's the plan for when we have a driver that doesn't use either UEFI or BIOS by default?

File: pkg/provisioner/ironic/ironic.go [hunk 6]
```diff
@@ -203,6 +209,12 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
 	}
 }
 
+// BootMode returns the boot method to be used for the host, based on
+// the BMC driver.
+func (p *ironicProvisioner) BootMode() v1alpha1.BootMode {
+	return p.bmcAccess.DefaultBootMode()
```
Comment (zaneb) [inline pkg/provisioner/ironic/ironic.go:215]: Do we need some way of getting the *allowed* boot modes?

Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go:215]: What would that be used for?

File: pkg/provisioner/ironic/ironic.go (hunk 5 above)
Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go:806]: Do you think we need to do more than extend the enum?

Comment (zaneb) [reply pkg/provisioner/ironic/ironic.go:806]: Depends on whether the right thing to do for Ironic is to send an empty string or not to send anything.

File: pkg/provisioner/ironic/ironic.go (hunk 6 above)
Comment (zaneb) [reply pkg/provisioner/ironic/ironic.go:215]: So we could ignore the user's preference if they require something Ironic can't support.
But maybe we just go ahead and try to provision and they get an error if they did it wrong.

Comment (juliakreger) [reply pkg/provisioner/ironic/ironic.go:215]: Well, provision should fail... that is unless the BMC is completely disjointed from reality. Unfortunately that is a thing, yet extremely rare and vendors I'm aware of having started to head down that path quickly reverted once people started complaining.

Realistically, if we're using IPMI, there is just no solid and consistent way to know. Some vendors are only settable via raw flags, others are only settable via custom boot target. Luckily everyone seems to support the raw flags, as far as I'm aware.

Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go:215]: > Well, provision should fail... that is unless the BMC is completely disjointed from reality. Unfortunately that is a thing, yet extremely rare and vendors I'm aware of having started to head down that path quickly reverted once people started complaining.
> 
> Realistically, if we're using IPMI, there is just no solid and consistent way to know. Some vendors are only settable via raw flags, others are only settable via custom boot target. Luckily everyone seems to support the raw flags, as far as I'm aware.

We've established that there is no way for us to know for sure, that's why we have the new field.

The only question in my mind is how to pick a default. Should we ask the BMC? Or should we use a single default all the time?

Comment (dhellmann) [reply pkg/provisioner/ironic/ironic.go:215]: Sounds good to me.

**Comment by stbenjam** [comment]
> I wonder if we want to remove the automatic setting stuff entirely and rely only on the field. Then we can set a default now, maybe use the hardware platform type to change that default later for ARM or whatever doesn't support the modes we have, and the user can always override the value.

I'm happy with this version, but are you thinking to remove all the implicit boot mode based on BMC? I think we were under the assumption that basically if you're using old IPMI protocols, you probably want old boot firmware, but that assumption was probably wrong.

I would be happy to set UEFI as a default for everyone, and let people override to use legacy/bios as needed.

**Comment by dhellmann** [comment]
/hold

See #602 instead

**Comment by dhellmann** [comment]
Closing in favor of #602

**Pull request closed without merging**